### PR TITLE
[#22] [Backend] As a user, I can see an article details

### LINKED ISSUE #22: As a user, I can see an article details
## Why
The users will be able to view a selected article details whether they are logged in or not with data provided from server.

## Acceptance Criteria
- [ ] Implement the API for getting an article details data via this [endpoint](https://github.com/gothinkster/realworld/tree/master/api#get-article).
- [ ] The endpoint takes the article's `slug` as the parameter in the request url.
- [ ] Create a usecase for getting an article details for later usage in integrate task.
- [ ] Parse the response into the corresponding model, here is an example response for a single article:
```
{
  "article": {
    "slug": "how-to-train-your-dragon",
    "title": "How to train your dragon",
    "description": "Ever wonder how?",
    "body": "It takes a Jacobian",
    "tagList": ["dragons", "training"],
    "createdAt": "2016-02-18T03:22:56.637Z",
    "updatedAt": "2016-02-18T03:48:35.824Z",
    "favorited": false,
    "favoritesCount": 0,
    "author": {
      "username": "jake",
      "bio": "I work at statefarm",
      "image": "https://i.stack.imgur.com/xHWG8.jpg",
      "following": false
    }
  }
}
```

### PR DESCRIPTION
**Pull request by markgravity**
Resolved #22

## What happened

Add Get Article API

## Insight

- Add `getArticle()` into `ArticleRepository
- Create `GetArticleUseCase`
- Rename `APIArticleResponse` to `APIArticlesResponse` to match with response which return multi articles
- Update `APIArticleResponse` to handle single article only

## Proof Of Work

Request Log
```
*** Overview *** 
URL https://conduit.productionready.io/api/articles/end-to-end-testing-with-cypress-wkyqoe
Method GET
Response Code 200
Request Start Time Sep 15 2021 - 09:03:25
Duration 14s

*** Request Header *** 
User-Agent Nimble Medium - Staging/1.0.1 (co.nimblehq.nimble-medium-stag; build:1; iOS 14.5.0) Alamofire/5.4.3
Accept-Language en;q=1.0
Accept-Encoding br;q=1.0, gzip;q=0.9, deflate;q=0.8


*** Request Body *** 
-

*** Response Header *** 
x-runtime 0.060466
x-content-type-options nosniff
x-powered-by Phusion Passenger 4.0.60
Server cloudflare
x-request-id c91f2447-a172-4ed2-be82-3c1ec1e20616
Etag W/"a50f3407969a299a0701fd627d3927fa"
nel {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
Cache-Control max-age=0, private, must-revalidate
x-frame-options SAMEORIGIN
Vary Origin
Content-Encoding br
cf-ray 68ee4791f9922ee6-SIN
Content-Type application/json; charset=utf-8
report-to {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=ZJ1VP809wKHVFaor3UxdXBQCBwgFqPUvGIHXB1GVRsJMrRqBlmVhaCvBY%2Bx335g2UV4vg809vATEd10ZiATk%2BCTesXKYN08waDlJDVvqhkQeJrf7%2B6Lqvy3VvT0OIQRVM1M3c0v%2BJC03A3QwDA%3D%3D"}],"group":"cf-nel","max_age":604800}
expect-ct max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
Date Wed, 15 Sep 2021 02:03:37 GMT
x-xss-protection 1; mode=block
Status 200 OK
cf-cache-status DYNAMIC
Alt-Svc h3=":443"; ma=86400, h3-29=":443"; ma=86400, h3-28=":443"; ma=86400, h3-27=":443"; ma=86400


*** Response Body *** 
{
  "article" : {
    "slug" : "end-to-end-testing-with-cypress-wkyqoe",
    "tagList" : [

    ],
    "author" : {
      "username" : "Kento",
      "image" : "https://static.productionready.io/images/smiley-cyrus.jpg",
      "following" : false,
      "bio" : null
    },
    "favoritesCount" : 0,
    "title" : "End to end testing with cypress",
    "favorited" : false,
    "updatedAt" : "2021-09-14T15:03:09.073Z",
    "createdAt" : "2021-09-14T15:03:09.073Z",
    "body" : "This is the body of the Article",
    "description" : "This is a description for an end to end testing using cypress"
  }
}

------------------------------------------------------------------------
------------------------------------------------------------------------
------------------------------------------------------------------------
```

Decode Log
```
2021-09-15 09:03:25.774: getArticle -> subscribed
2021-09-15 09:03:40.420: getArticle -> Event next(DecodableArticle(slug: "end-to-end-testing-with-cypress-wkyqoe", title: "End to end testing with cypress", description: "This is a description for an end to end testing using cypress", body: "This is the body of the Article", tagList: [], createdAt: 2021-09-14 15:03:09 +0000, updatedAt: 2021-09-14 15:03:09 +0000, favorited: false, favoritesCount: 0, apiAuthor: NimbleMedium.DecodableProfile(username: "Kento", bio: nil, image: Optional("https://static.productionready.io/images/smiley-cyrus.jpg"), _following: DefaultCodable.Default<DefaultCodable.False>(wrappedValue: false))))
2021-09-15 09:03:40.425: getArticle -> Event completed
2021-09-15 09:03:40.425: getArticle -> isDisposed
```

### CODE REVIEW

File: NimbleMediumTests/Sources/Specs/Data/Repositories/ArticleRepositorySpec.swift [hunk 1]
```diff
@@ -87,6 +87,52 @@ final class ArticleRepositorySpec: QuickSpec {
...
+                    beforeEach {
+                        outputArticle = scheduler.createObserver(DecodableArticle.self)
+                        networkAPI.setPerformRequestForReturnValue(Single.just(inputResponse))
+                        repository.getArticle(slug: "slug")
+                        .asObservable()
+                        .compactMap {
+                            $0 as? DecodableArticle
+                        }
+                        .bind(to: outputArticle)
+                        .disposed(by: disposeBag)
```
Comment (minhnimble) [inline NimbleMediumTests/Sources/Specs/Data/Repositories/ArticleRepositorySpec.swift:107]: the indentation looks a bit weird here, please help check again 🙏

File: NimbleMediumTests/Sources/Specs/Data/Repositories/ArticleRepositorySpec.swift [hunk 2]
```diff
@@ -87,6 +87,52 @@ final class ArticleRepositorySpec: QuickSpec {
...
+
+                    beforeEach {
+                        outputError = scheduler.createObserver(Optional<Error>.self)
+                        networkAPI.setPerformRequestForReturnValue(Single<APIArticleResponse>.error(TestError.mock))
+                        repository.getArticle(slug: "slug")
+                        .asObservable()
+                        .materialize()
+                        .map { $0.error }
+                        .bind(to: outputError)
+                        .disposed(by: disposeBag)
```
Comment (minhnimble) [inline NimbleMediumTests/Sources/Specs/Data/Repositories/ArticleRepositorySpec.swift:127]: same here

**Review by minhnimble — APPROVED** [review]
LGTM 🚀

File: NimbleMedium/Sources/Data/NetworkAPI/RequestConfigurations/ArticleRequestConfiguration.swift [hunk 3]
```diff
@@ -53,6 +56,8 @@ extension ArticleRequestConfiguration: RequestConfiguration {
                 "offset": offset
             ]
             return parameters.compactMapValues { $0 }
+        case .getArticle:
+            return [:]
```
Comment (suho) [inline NimbleMedium/Sources/Data/NetworkAPI/RequestConfigurations/ArticleRequestConfiguration.swift:60]: We can return nil

**Review by suho — APPROVED** [review]

**Pull request merged**